### DISCUSSION TITLE
MVC API Design Best Practice for Query Params and HTTP Post Body

### DISCUSSION THREAD

**Discussion by saikarthikp9** [prisma/prisma#17599] [Q&A]
Why do we follow some other standards for the API endpoint design and then convert the incoming query params (think for filtering, etc.) and/or request body to the Prisma format to make the query? Can't we just avoid that step and design the API endpoint such that it can almost exactly be used with Prisma? I am yet to find a project online that does this though. 

A simple example of a Nodejs MVC app: 
route.ts:
```
import * as controller from '../../controllers/testController';
router.get('/', controller.getAllTests);
router.post('/', controller.createNewTest);
router.delete('/:id', controller.deleteOneTest);
```
testController.ts:
```
import * as service from '../services/testService';

export const getAllTests = async (req: Request, res: Response) => {
    const query = qs.parse(req.query);
    const allTests = await service.getAllTests(query);
    res.send(allTests);
};
export const createNewTest = (req: Request, res: Response) => {
    const createdTest = service.createNewTest(req.body);
    res.send(createdTest);
};
export const deleteOneTest = async (req: Request, res: Response) => {
    const test = await service.deleteOneTest(
        req.params.id
    );
    res.send(test);
};

```
testService.ts:
```
import { Prisma, PrismaClient as prisma } from '../utils/PrismaClient';

export const getAllTests = async (
    query: Prisma.TestFindManyArgs
) => {
    const tests = await prisma.test.findMany(query);
    return tests;
};

export const getOneTest = async (id: string) => {
    const query: Prisma.TestFindFirstArgsBase = {
        where: {
            id
        }
    };
    const test = await prisma.test.findFirst(query);
    if (test == null) throw new NotFoundError();
    return test;
};

export const createNewTest = async (
    data: Prisma.TestCreateInput
) => {
    const test = await prisma.test.create({ data });
    return test;
};


export const deleteOneTest = async (id: string) => {
    const test = await prisma.test.delete({
        where: {
            id
        }
    });
    return test;
};

```

And then making a request like: 
```
https://test.com/api/v1/restaurants?where[name]=OliveGarden
```

**Comment by nurul3101** [comment]
Hello @saikarthikp9 👋 

You can design your API endpoints to match the Prisma format. But, Ideally, the API should have its own standard and design, and this standard should be consistent across all the different APIs that you may create in your application.

By having a standardized way of designing your APIs, you can maintain consistency in how the APIs are defined. Prisma doesn't enforce using only one particular standard to design the APIs, you can select the one that works best for you.

**Reply by saikarthikp9** [reply]
Do you see any issues with following the Prisma format across all APIs? I don't really foresee any issues or see any reason not to follow it. If in the future, I need to create an API that does NOT use Prisma upstream, I can just do a conversion of the query params/body to fit the requirement of the upstream call.

**Reply by nurul3101** [reply]
I don't see any issues. If your business requirements are satisfied, you are free to use any standard that you are comfortable with.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
